From c21d0792b054dd366d08f4dfac2ac121bbbbe65b Mon Sep 17 00:00:00 2001
From: Miquel Raynal <miquel.raynal@bootlin.com>
Date: Tue, 24 Dec 2024 18:05:57 +0100
Subject: [PATCH 02248/12543] spi: nxp-fspi: Support per spi-mem operation
 frequency switches

[ Upstream commit 26851cf65ffca2d3a8d529a125e54cf0084d69e7 ]

Every ->exec_op() call correctly configures the spi bus speed to the
maximum allowed frequency for the memory using the constant spi default
parameter. Since we can now have per-operation constraints, let's use
the value that comes from the spi-mem operation structure instead. In
case there is no specific limitation for this operation, the default spi
device value will be given anyway.

The per-operation frequency capability is thus advertised to the spi-mem
core.

Cc: AngeloGioacchino Del Regno <angelogioacchino.delregno@collabora.com>
Signed-off-by: Miquel Raynal <miquel.raynal@bootlin.com>
Link: https://patch.msgid.link/20241224-winbond-6-11-rc1-quad-support-v2-12-ad218dbc406f@bootlin.com
Signed-off-by: Mark Brown <broonie@kernel.org>
Stable-dep-of: 40ad64ac25bb ("spi: nxp-fspi: Propagate fwnode in ACPI case as well")
Signed-off-by: Sasha Levin <sashal@kernel.org>
---
 drivers/spi/spi-nxp-fspi.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/spi/spi-nxp-fspi.c b/drivers/spi/spi-nxp-fspi.c
index 9162e60e01b8..3a4cc7357648 100644
--- a/drivers/spi/spi-nxp-fspi.c
+++ b/drivers/spi/spi-nxp-fspi.c
@@ -818,7 +818,7 @@ static void nxp_fspi_select_mem(struct nxp_fspi *f, struct spi_device *spi,
 	dev_dbg(f->dev, "Target device [CS:%x] selected\n", spi_get_chipselect(spi, 0));
 
 	nxp_fspi_select_rx_sample_clk_source(f, op);
-	rate = min(f->support_max_rate, spi->max_speed_hz);
+	rate = min(f->support_max_rate, op->max_freq);
 
 	if (!op->cmd.dtr) {
 		f->flags &= ~FSPI_DTR_MODE;
@@ -1328,10 +1328,12 @@ static const struct spi_controller_mem_ops nxp_fspi_mem_ops = {
 
 static const struct spi_controller_mem_caps nxp_fspi_mem_caps = {
 	.dtr = true,
+	.per_op_freq = true,
 };
 
 static const struct spi_controller_mem_caps nxp_fspi_mem_caps_quirks = {
 	.dtr = false,
+	.per_op_freq = true,
 };
 
 static void nxp_fspi_cleanup(void *data)
-- 
2.53.0

